pkg/cloudprovider/providers/openstack/openstack_instances.go

```diff
@@ -132,7 +136,24 @@ func (i *Instances) InstanceExistsByProviderID(ctx context.Context, providerID s
 
 // InstanceShutdownByProviderID returns true if the instances is in safe state to detach volumes
 func (i *Instances) InstanceShutdownByProviderID(ctx context.Context, providerID string) (bool, error) {
-	return false, cloudprovider.NotImplemented
+	instanceID, err := instanceIDFromProviderID(providerID)
+	if err != nil {
+		return false, err
+	}
+
+	server, err := servers.Get(i.compute, instanceID).Extract()
+	if err != nil {
+		if isNotFound(err) {
+			return false, nil
```

Comment:
Why not return NotFound err here

Reply:
To avoid misleading log print https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L257-L259 the notfound thing is checked again in row https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L274

These two calls could be combined but its not that easy when we have many cloudproviders and things are not working in similar way.

Reply:
The NotFound error message is printed twice, i don't think that is misleading, they just meet the same error. 
Since the NotFound error is still a kind of error, i'd prefer returning that...

Reply:
+1 @NickrenREN